### ISSUE TITLE
uploadDir should get complete path passed to filter

### ISSUE THREAD

**Issue by jonnytest1** [theophilusx/ssh2-sftp-client#383]
the filter in uploadDir only gets the current file name but getting the complete path passed would be helpful 
(to keep it backwards compatible you could also add the option to pass a function instead)

(
PS if possible i think using 
Promise.all(dirEntries.map(async entry=>{do recursive stuff...})) 
could have better runtime since stuff can be done in parrallel - not sure if the connection may get screwed up tho
)

**Comment by theophilusx** [comment]
I will need to consider that change - there are some other changes I'm
already considering.

Note however, array comp functions like map, foreEach, filter etc are
not async safe. While you can use them with async, there are additional
steps you have to take. Generally, I find just using a for loop is
cleaner and clearer.

The other issue is that some sftp servers seem better at handling lots
of concurrent transfers than others. My testing has also found that
often, you don't get the increased throughput you would expect, though a
lot can depend on the size of your files. In many cases, the context
switching can offset any benefit from trying to process 'parallel'
transfers (largely due to the fact, node/javascript is single threaded
and what your really doing is 'time slicing' rather than real parallel
processing)

In testing, I found you only got a very small performance boost doing
parallel requests when the files were large (Gb size) and even then, it
only took a very few number of parallel requests before at least one
segment on your network path would reach capacity (which might be due to
router rate limiting and/or server rate limiting). Once you reach that
point, throughput is fixed and adding additional transfer tasks begins
to degrade performance as more context switching occurs.

**Comment by jonnytest1** [comment]
ill leave the decision making to you ^^ you deifnitely seem to know ur stuff ;)
but since i already added a version for my local setup (i needed to transform the files as well )
im just gonna drop the code here :

// it definitely seemed faster than the current version oO
 ```typescript
   async getFileHandle() {
        if (this.handleCount < 1000) {
            return
        }
        await new Promise(res => setTimeout(res, 1))
        return this.getFileHandle()
    }
 
 async uploadDirRecursive(from: string, to: string) {
        await this.readyPromise
        const entries = await readdir(from, { withFileTypes: true })
        await Promise.all(entries.map(async entry => {
            const fromPAth = win32.join(from, entry.name)
            const remotePath = posix.join(to, entry.name)
            if (shouldIgnoreFile(remotePath)) {       // doing the filters
                return
            }
            if (entry.isDirectory()) {
                await this.mkdir(remotePath)
                await this.uploadDirRecursive(fromPAth, remotePath)
            } else {
                await this.copy(fromPAth, remotePath, false)
            }
        }))
    }
     async copy(from: string, to: string, changeCheck = true) {

        try {
            if (changeCheck) {
               // i did some timestatamp change check here (probably faster to just copy it ^^ )
            }
            console.log(`copying ${from} to ${to}`)
            await this.readyPromise   // waiting for connection finish 
            await this.getFileHandle() // make sure to not do toooo many requests
            this.handleCount++
            let pr: string
            if (this.isTextual(from)) {
               // my transformation
                const file = await readFile(from, { encoding: "utf-8" })
                const linuxFile = file.replace(/\r\n/g, "\n")
                const linuxFileBfr = Buffer.from(linuxFile);

                pr = await this.client.put(linuxFileBfr, to)
            } else {
                pr = await this.client.put(from, to)
            }
            this.handleCount--
        } catch (e) {
            debugger;
        }
    }
    
    
```

**Comment by theophilusx** [comment]
I have now rfeworked the uploadDir and downloadDir functions so that they now accept a predicate function which will determine whether an item is to be uploaded/downloaded. The function will be called with two arguments - 1st argument is the full path to the item and the second argument is a boolean, which will be true if the item is a directory. The predicate function should return true to have the item included in the upload/download and false to have it excluded. 

This new form of these two methods will be included in v8.0.0, which should be released in the next week or so (just waiting on some testing on some changes in other areas for v8.0.0). Closing this issue now.
